providers/implementations/kdfs/tls1_prf.c

```diff
@@ -107,6 +107,11 @@ typedef struct {
     /* Concatenated seed data */
     unsigned char *seed;
     size_t seedlen;
+
+#ifdef FIPS_MODULE
+    PROV_DIGEST digest;
```

Comment:
Why do you need to do this for FIPS?

Reply:
To determine whether the underlying digest function is approved, it is required to retrieve related information from `P_hash` (EVP_MAC_CTX).

However, there is no gettable parameter called `OSSL_MAC_PARAM_DIGEST` in EVP_MAC or EVP_MAC_CTX, so I store the used digest information separately here.

Reply:
The digest will be fetched from the FIPS provider's internal library context.  There is no possibility of it being unapproved.  At least until there are unapproved digests inside the FIPS boundary.

Limiting the types of digest used to those mandated by TLS seems like an independent question to their FIPSness.  I.e. if you used SHA 256/224, TLS isn't going to work but I'm not sure it would be unapproved in a FIPS sense.  I haven't gone to the standards to check if there is a special restriction in this case or not.

Reply:
The reason I add the digest check for TLS1-PRF (or TLS13) is that I think it can be treated a standalone service (to derive a key no matter if it is used in TLS. of course, the possibility of this usage may be very small).

Therefore, we need a indicator to tell the user whether the operation is approved or not.

Reply:
It simply is not possible to get a non-approved digest here, so I don't see a need to check for that case.
Using this PRF outside of TLS is almost certainly unapproved, which is impossible to detect.

Does a NIST standard explicitly limit the digests that can be used with this PRF?
I don't remember seeing such a limitation.

TLS does restrict the usable digests, but that's not a NIST mandate.

Reply:
A quick check of SP 800-135r1 reveals: _The TLS 1.0 and 1.1 KDF is performed in the context of the TLS protocol_ and _The TLS 1.2 KDF is performed in the context of the TLS protocol_.

The non-TLS use case is forbidden, so we don't need this check.

Reply:
However, SP 800-135r1 mentions <code>The TLS 1.0 and 1.1 KDF is <b>approved</b> when the following conditions are satisfied</code>, so it should report the operation is **unapproved** if we use the KDF outside the context of the TLS protocol or use unlisted digest in the KDF ?

Reply:
We cannot tell if it's usage is or is not part of TLS.

Reply:
Please allow me confirm one more question.

So, if we follow this thought, these KDFs (i.e. SSH, X9.63, TLS1.x KDFs) should still have an indicator but always report the operation is unapproved ?

Reply:
@slontis About digest check for KDFs (e.g. SSH, TLS) listed in SP 800-135r1, do you have an opinion on this ?

Reply:
I would assume these algorithms are being used by the intended protocol,
and only allow digests that can be tested (so match whatever the ACVP server allows) - Which hopefully correspond to SP 800-135r1.

Reply:
i.e. They are approved.

Reply:
How the code is used in a protocol is outside the scope of the FIPS module.

Reply:
I think what Pauli was saying is that the the fips module only allows digests that come from the fipsprovider. And that list is the following... (See fipsprov.c)
SHA1,SHA224,SHA256,SHA384,SHA512,SHA512_224,SHA512_256,SHA3_224,SHA3_256,SHA3_384,SHA3_512,SHAKE128, SHAKE256.

Reply:
If the ACVP tests only allow a subset of this then that is what we should limit the digest checks to.

Reply:
@paulidale part of the design doc relates to what algorithms are testable. If they are not listed on the cert then they are not approved. (And they are only listed if they have been tested)

Reply:
> If the ACVP tests only allow a subset of this then that is what we should limit the digest checks to.

If so, I think we should keep all digest checks in this PR because the ACVP test for these KDFs (except HKDF and SSKDF) only allow a subset of the digests included in FIPS provider.